link-grammar/parse/preparation.c

```diff
@@ -160,21 +174,37 @@ static void print_connector_list(const char *s, const char *t, Connector * e)
  * as strings, which are used for getting a unique identifier for the
  * starting connector of each such sequence.
  * In order to save the need to cache the endpoint word numbers the
- * connector identifiers are not shared between words. To that end the
- * word number is prepended to the said strings.
+ * connector identifiers should not be shared between words. The initial
+ * implementation tried to do that by prepending the word number to the
+ * connector list. However, this is buggy, because trailing sequences
+ * should be unique per word. But this is fixed anyway by the alternatives
+ * check described below. (This bug left undetected because the hash
+ * tables are now relatively empty and also we still hash with word
+ * numbers - for added entropy).
```

Comment:
So did you test, by making the hash tables small, so that they are dense?  If your theory is correct, the bug should then become visible.

Reply:
I said this only by analyzing how the "trailing connectors hash" works.
Since the word numbers are not checked on cache reads, their result depends solely on the endpoint hash values. On my initial implementation they were shared (unintentionally) by the same trailing sequences of deep connectors on different words, but each should have its own independent number of linkages.
The problem for different words can be solved by checking the words numbers too - at runtime -as was done originally. However, it is more efficient to one-time encode a different suffix_id number instead. The problem of tokens of different alternatives on the same word slot, that also clearly cannot share the exact same number of linkages, can only be solved by a different suffix_id
So a different suffix_id per Gword (which means a different word or a different alternative) solves both problem.

Just now I tried the current repository version with one-slot cache size (shift=0). Surprisingly, `ru` then runs only a few percents slower. But it gives the same 47 errors. However, when producing all the linkages, one sentence (the last one) gives a different result, which is strange:
`Found 2147483647 linkages (0 of 0 random linkages had no P.P. violations) at null count 5
`
Note only the `0 of 0` is strange, but this number and the null-count is different on every run!
`Found 565524126 linkages (0 of 0 random linkages had no P.P. violations) at null count 2
`
I tried it with no address randomization (a mode I use for debug) and it is still non-deterministic (only for this sentence). This needs a deeper investigation. With the default cache size it is deterministic:
`Found 2376 linkages (1000 of 1000 random linkages had no P.P. violations) at null count 7
`

I also tried the fixes batch with 1 hash slot. It runs about 50x slower...
One sentence , that get parsed with the default cache size, doesn't get parsed then (however it get parsed fine when this PR is applied):
```
linkparser> People ages 55 to 64 on average are out of work for more than 54 weeks, while those ages 25 to 34 are unemployed for about 36 weeks, according to figures from the Bureau of Labor Statistics.
...
Found 1745111394 linkages (0 of 0 random linkages had no P.P. violations) at null count 2
```
This is a similar problem to the one is `ru`, including the random number of linkages at a random null-count on each try... BTW, this sentence has no alternatives. So this may be due to the word number encoding problem (bug (1) in the initial message).

 Like the `ru` problem, this PR also fixes this problem.

The `he` and `lt` batches give the exact same detailed linkage results when 1-slot is used.

Beside of these 2 problems there are no apparent wrong results in the current repository version. But I already know from a large amount of Connector_table hacking that subtle caching bugs may manifest themselves very rarely, sometimes in a single sentence.

Because these 2 problems are fixed in this PR, and the theoretical considerations that see to me right, **I propose to accept this PR** even so it causes a slight slowness on long sentences. Of course I will think how to improve that, as this whole idea (of trailing connectors hashing) is only several days old (and seems to have more theoretical implications).

I will also investigate this strange `0 of 0 random linkages` problem (even though it will not exist if this PR is applied) because it may hint on something fundamentally wrong.